server/src/main/scala/org/http4s/server/middleware/Timeout.scala

```diff
@@ -87,13 +88,15 @@ object Timeout {
     * Internal Server Error` response
     * @param service [[HttpRoutes]] to transform
     */
-  def apply[F[_]](timeout: FiniteDuration, timeoutResponse: F[Response[F]])(
-      @deprecatedName('service) routes: HttpRoutes[F])(
-      implicit F: Concurrent[F],
-      T: Timer[F]): HttpRoutes[F] = {
-    val OTC = Concurrent[OptionT[F, ?]]
-    routes
-      .mapF(respF => OTC.race(respF, OptionT.liftF(T.sleep(timeout) *> timeoutResponse)))
+  def apply[F[_], G[_], A](timeout: FiniteDuration, timeoutResponse: F[Response[G]])(
+      @deprecatedName('service) http: Kleisli[F, A, Response[G]])(
+      implicit F: Concurrent[F]): Kleisli[F, A, Response[G]] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    val T = Timer.derive[F]
+
```

Comment:
I was seeing the following error:

```
[error] /Users/tom/Projects/http4s/examples/blaze/src/main/scala/com/example/http4s/blaze/demo/server/Module.scala:48:22: Cannot find implicit value for Timer[[A]cats.data.OptionT[F,A]].
[error] Note that [A]cats.data.OptionT[F,A] needs to be a cats.effect.Async data type. You might also
[error] need a scala.concurrent.ExecutionContext in scope, or equivalent, try to
[error] import scala.concurrent.ExecutionContext.Implicits.global
[error]     Timeout(1.second)(timeoutHttpEndpoint)
```

We already require `Concurrent[F]`, which implies `LiftIO[F]`, so we're able to `Timer.derive[F]` where we need it.

If we're happy to do this, is it OK to use the global `ExecutionContext` here, or should we require a `Strategy`/`ExecutionStrategy` be passed in?

Reply:
I would prefer you simply take a `Timer` instead of calling `derive` since it opens feeding in a different one if need be.

Reply:
Right! And on the issue of there not being an instance for `OptionT[F, ?]` - do you have any suggestions on how best to provide one, or would we leave this to users?

Reply:
that's in userland/another PR I think.

Reply:
I get you. I'm happy to defer to your judgement on this, but would say that:

1. I think I'd argue that it's an implementation detail of the `Timeout` middleware and it's not immediately obvious to me why I might want to switch one `Timer` out for another
2. I think the ergonomics are better this way: it's cleaned up `Module` and `Server` in this PR and I think I prefer this API as a user

If that's not convincing, that's fine, I'll change it 😄

Reply:
Counter args:

1. the global ec is often used as the general client pool in something like blaze. Where a user chooses to execute stuff is _not_ an implementation detail in my eyes. It impacts the app, even if it's true that a good chunk of the times  people do use global. Whenever you `Timer.sleep`, you force the continuation of that call onto the other pool provided by the `timer`. For advanced users that _do_ configure thread pools, this is a net loss.
2. What does it clean up aside from removing a single implicit? 

I'm in general :-1: on explicitly less powerful configuration. It's a slight win for beginner users by limiting advanced users by forcing them onto `global`. Moreover, I don't particularly like considering application configuration (which timer propagates, essentially since it does carry a thread pool reference no matter how much anyone wants to pretend it doesn't) "an implementation detail".

This is a democratic decision though, so other maintainer opinions would be nice.

Reply:
I don't disagree with any of that! To be clear, I'm only arguing for the `Timer` to be defined locally. The `ExecutionContext` can absolutely be passed in.

I'd also welcome other POVs. I'm acutely aware that I'm arguing from a position of ignorance on this!

Reply:
but that's the thing though: in the ecosystem we're trying to move away from passing `ExecutonContext`s and passing `Timer[F]`s. Hence my quip.

Reply:
I see! I'll change it.